File: muoni/C++/src/Values/FloatValue.cpp
//
// Created by Joel Courtney on 2019-03-01.
//

#include "Values/FloatValue.h"
#include "Values/ValueFactory.h"
#include "Types/IntType.h"

FloatValue::FloatValue(double d) : d(d) {}

std::string FloatValue::toString() const {
    return std::to_string(d);
}

Type* FloatValue::getType() const {
    return IntType::getInstance();
}

void FloatValue::increment() {
    d++;
}

void FloatValue::decrement() {
    d--;
}

void FloatValue::negate() {
    d = -d;
}

void FloatValue::conjugate() {}

ScalarValue* FloatValue::magnitude() {
    if (d >= 0) {
        return ValueFactory::from(d);
    } else {
        return ValueFactory::from(-d);
    }
}

File: muoni/Java/src/Values/FloatValue.java
package values;

public class FloatValue implements ScalarValue {
	private double d;

	public FloatValue(double _d) {
		d = _d;
	}

	public void increment() {
		d++;
	}
	public void decrement() {
		d--;
	}

	public void negate() {
		d = -d;
	}
	public FloatValue magnitude() {
		if (d >= 0) {
			return new FloatValue(-d);
		} else {
			return new FloatValue(d);
		}
	}
	public void conjugate() {}

	public String toString() {
		return Double.toString(d);
	}
}
